### PR TITLE
bpo-30038: fix race condition in signal delivery + wakeup fd

### PR DESCRIPTION
**Pull request by njsmith**
Before, it was possible to get the following sequence of
events (especially on Windows, where the C-level signal handler for
SIGINT is run in a separate thread):

- SIGINT arrives
- trip_signal is called
- trip_signal writes to the wakeup fd
- the main thread wakes up from select()-or-equivalent
- the main thread checks for pending signals, but doesn't see any
- the main thread drains the wakeup fd
- the main thread goes back to sleep
- trip_signal sets is_tripped=1 and calls Py_AddPendingCall to notify
  the main thread the it should run the Python-level signal handler
- the main thread doesn't notice because it's asleep

This has been causing repeated failures in the Trio test suite:
  https://github.com/python-trio/trio/issues/119

### CODE REVIEW

**Comment by njsmith** [comment]
@haypo may also be relevant, as AFAICT he was the main author of `trip_signal` and the wakeup fd functionality

**Comment by njsmith** [comment]
Maybe this will be a useful workaround for the test suite until fixed versions of CPython are released? https://github.com/box/flaky

**Comment by njsmith** [comment]
doh, posted to the wrong issue, sorry

**Comment by njsmith** [comment]
3 week ping.

This sounds complicated, but it's pretty straightforward really: if you're using a condition variable, e.g. to important a queue, you set the new value *then* signal the condition to wake up anyone who was waiting. If you do it in the other order, your queue might deadlock. Similarly, here we need to set the variable letting the main thread know that a signal has arrived *before* we wake it up. The patch is just swapping the order of these two things.

Unfortunately, like any race condition, it's difficult to test, but if anyone has a Windows build setup then there's a reproducer in the linked bpo issue. And even if not, the current approach is obviously broken by inspection.

File: Modules/signalmodule.c [hunk 1]
```diff
@@ -244,6 +244,16 @@ trip_signal(int sig_num)
...
 
+    if (!is_tripped) {
+        /* Set is_tripped after setting .tripped, as it gets
+           cleared in PyErr_CheckSignals() before .tripped. */
+        is_tripped = 1;
+        Py_AddPendingCall(checksignals_witharg, NULL);
+    }
+
+    /* And then write to the wakeup fd *after* setting all the globals and
+       doing the Py_AddPendingCall (bpo-30038) */
```
Comment (vstinner) [inline Modules/signalmodule.c]: I would like to see here the rationale for setting is_tripped before writing into the wake up fd. You already wrote it in the http://bugs.python.org/issue30038 and in the PR description. Just copy it here please ;-)

Comment (njsmith) [reply Modules/signalmodule.c]: Done

**Pull request merged**